File: winai/services/api/lead.service.ts
/**
 * Lead Service - Integração com API de Leads/CRM
 */

import { httpClient } from './http-client';

// ============================================
// Types
// ============================================

export type LeadStatusType =
    | 'NEW'
    | 'CONTACTED'
    | 'QUALIFIED'
    | 'MEETING_SCHEDULED'
    | 'PROPOSAL_SENT'
    | 'NEGOTIATION'
    | 'WON'
    | 'LOST';

export interface LeadTagData {
    id: string;
    name: string;
}

export interface LeadData {
    id: string;
    name: string;
    email: string;
    phone: string | null;
    status: LeadStatusType;
    statusLabel: string;
    ownerName: string | null;
    notes: string | null;
    source: string | null;
    trackId?: string | null;
    trackSource?: string | null;
    utmSource?: string | null;
    utmMedium?: string | null;
    utmCampaign?: string | null;
    utmContent?: string | null;
    utmTerm?: string | null;
    gclid?: string | null;
    fbclid?: string | null;
    estimatedValue?: number | null;
    leadScore?: number | null;
    /** Etiquetas CRM da empresa (segmentos, carteiras, produtos…). */
    tags?: LeadTagData[];
    createdAt: string;
    updatedAt: string;
}

export interface LeadRequest {
    name: string;
    email: string;
    phone?: string;
    status?: LeadStatusType;
    ownerName?: string;
    notes?: string;
    source?: string;
    trackId?: string;
    trackSource?: string;
    utmSource?: string;
    utmMedium?: string;
    utmCampaign?: string;
    utmContent?: string;
    utmTerm?: string;
    gclid?: string;
    fbclid?: string;
    estimatedValue?: number | null;
    leadScore?: number | null;
    /** Nomes das etiquetas CRM. Enviar lista vazia remove todas; omitir em atualizações parciais se o backend aceitar. */
    tags?: string[];
}

export interface PagedResponse<T> {
    content: T[];
    totalElements: number;
    totalPages: number;
    size: number;
    number: number;
    first: boolean;
    last: boolean;
}

/** Labels alinhados ao pipeline AMPLIA / Kanban */
export const LEAD_STATUS_LABELS: Record<LeadStatusType, string> = {
    NEW: 'Novos Leads',
    CONTACTED: 'Em Contato',
    QUALIFIED: 'Qualificados',
    MEETING_SCHEDULED: 'Reunião',
    PROPOSAL_SENT: 'Proposta',
    NEGOTIATION: 'Negociação',
    WON: 'Ganhos',
    LOST: 'Perdidos',
};

export const LEAD_STATUS_STYLES: Record<LeadStatusType, string> = {
    NEW: 'bg-blue-50 text-blue-600 border-blue-100',
    CONTACTED: 'bg-amber-50 text-amber-600 border-amber-100',
    QUALIFIED: 'bg-purple-50 text-purple-600 border-purple-100',
    MEETING_SCHEDULED: 'bg-indigo-50 text-indigo-600 border-indigo-100',
    PROPOSAL_SENT: 'bg-cyan-50 text-cyan-600 border-cyan-100',
    NEGOTIATION: 'bg-orange-50 text-orange-600 border-orange-100',
    WON: 'bg-emerald-500 text-white border-transparent',
    LOST: 'bg-rose-50 text-rose-600 border-rose-100',
};

export const KANBAN_COLUMN_ORDER: LeadStatusType[] = [
    'NEW',
    'CONTACTED',
    'QUALIFIED',
    'MEETING_SCHEDULED',
    'PROPOSAL_SENT',
    'NEGOTIATION',
    'WON',
    'LOST',
];

export const KANBAN_COLUMN_COLORS: Record<LeadStatusType, string> = {
    NEW: 'bg-blue-500',
    CONTACTED: 'bg-amber-500',
    QUALIFIED: 'bg-purple-500',
    MEETING_SCHEDULED: 'bg-indigo-500',
    PROPOSAL_SENT: 'bg-cyan-500',
    NEGOTIATION: 'bg-orange-500',
    WON: 'bg-emerald-600',
    LOST: 'bg-rose-500',
};

function normalizeLeadTags(raw: unknown): LeadTagData[] | undefined {
    if (raw == null) return undefined;
    const arr = Array.isArray(raw) ? raw : (raw as { tags?: unknown }).tags;
    if (!Array.isArray(arr)) return [];
    const out: LeadTagData[] = [];
    for (const item of arr) {
        if (item == null || typeof item !== 'object') continue;
        const o = item as Record<string, unknown>;
        const id = o.id != null ? String(o.id) : '';
        const name = o.name != null ? String(o.name).trim() : '';
        if (name) out.push({ id, name });
    }
    return out;
}

/** Garante número em estimatedValue (API pode enviar string ou snake_case em alguns ambientes). */
export function normalizeLeadData(raw: LeadData): LeadData {
    const r = raw as LeadData & { estimated_value?: unknown; tags?: unknown };
    const candidate = r.estimatedValue ?? r.estimated_value;
    const tags = normalizeLeadTags(r.tags);
    let estimatedValue: number | null | undefined = raw.estimatedValue ?? null;
    if (candidate === undefined || candidate === null || candidate === '') {
        estimatedValue = raw.estimatedValue ?? null;
    } else {
        const n = typeof candidate === 'number' ? candidate : Number(candidate);
        estimatedValue = Number.isNaN(n) ? raw.estimatedValue ?? null : n;
    }
    return { ...raw, estimatedValue, tags: tags ?? undefined };
}

// ============================================
// Service
// ============================================

/** Títulos customizados das colunas do Kanban (empresa) — chaves = LeadStatusType. */
export type CrmKanbanColumnTitles = Partial<Record<LeadStatusType, string>>;

function normalizeKanbanTitlesPayload(raw: Record<string, string> | undefined | null): CrmKanbanColumnTitles {
    const out: CrmKanbanColumnTitles = {};
    if (!raw) return out;
    for (const [k, v] of Object.entries(raw)) {
        if (KANBAN_COLUMN_ORDER.includes(k as LeadStatusType) && typeof v === 'string' && v.trim()) {
            out[k as LeadStatusType] = v.trim();
        }
    }
    return out;
}

export const leadService = {
    async getKanbanColumnTitles(): Promise<CrmKanbanColumnTitles> {
        const r = await httpClient.get<{ titles: Record<string, string> }>('/leads/kanban-column-titles');
        return normalizeKanbanTitlesPayload(r.titles);
    },

    async updateKanbanColumnTitles(titles: CrmKanbanColumnTitles): Promise<CrmKanbanColumnTitles> {
        const r = await httpClient.put<{ titles: Record<string, string> }>('/leads/kanban-column-titles', {
            titles,
        });
        return normalizeKanbanTitlesPayload(r.titles);
    },

    async getCrmTags(): Promise<LeadTagData[]> {
        const data = await httpClient.get<LeadTagData[]>('/leads/tags');
        return (data || []).map((t) => ({ id: String(t.id), name: String(t.name || '').trim() })).filter((t) => t.name);
    },

    async getAllLeads(): Promise<LeadData[]> {
        const data = await httpClient.get<LeadData[]>('/leads');
        return data.map(normalizeLeadData);
    },

    async getLeadsPaged(page: number = 0, size: number = 20): Promise<PagedResponse<LeadData>> {
        const p = await httpClient.get<PagedResponse<LeadData>>(`/leads/paged?page=${page}&size=${size}`);
        return { ...p, content: p.content.map(normalizeLeadData) };
    },

    async searchLeads(query: string, page: number = 0, size: number = 20): Promise<PagedResponse<LeadData>> {
        const p = await httpClient.get<PagedResponse<LeadData>>(
            `/leads/search?q=${encodeURIComponent(query)}&page=${page}&size=${size}`
        );
        return { ...p, content: p.content.map(normalizeLeadData) };
    },

    async getLeadById(id: string): Promise<LeadData> {
        return normalizeLeadData(await httpClient.get<LeadData>(`/leads/${id}`));
    },

    async createLead(lead: LeadRequest): Promise<LeadData> {
        return normalizeLeadData(await httpClient.post<LeadData>('/leads', lead));
    },

    async updateLead(id: string, lead: LeadRequest): Promise<LeadData> {
        return normalizeLeadData(await httpClient.put<LeadData>(`/leads/${id}`, lead));
    },

    async deleteLead(id: string): Promise<void> {
        await httpClient.delete(`/leads/${id}`);
    },
};

export default leadService;

File: winai-backend/src/main/java/com/backend/winai/service/LeadService.java
package com.backend.winai.service;

import com.backend.winai.dto.request.LeadRequest;
import com.backend.winai.dto.response.LeadResponse;
import com.backend.winai.dto.response.LeadTagResponse;
import com.backend.winai.entity.Company;
import com.backend.winai.entity.CrmLeadTag;
import com.backend.winai.entity.Lead;
import com.backend.winai.entity.LeadStatus;
import com.backend.winai.entity.WhatsAppConversation;
import com.backend.winai.repository.CompanyRepository;
import com.backend.winai.repository.CrmLeadTagRepository;
import com.backend.winai.repository.LeadRepository;
import com.backend.winai.repository.WhatsAppMessageRepository;
import com.backend.winai.repository.WhatsAppConversationRepository;
import com.backend.winai.repository.MeetingRepository;
import com.backend.winai.repository.WhatsAppBroadcastRecipientRepository;
import com.backend.winai.repository.FollowUpStatusRepository;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class LeadService {

    private static final int CRM_KANBAN_TITLE_MAX_LEN = 40;

    private final LeadRepository leadRepository;
    private final CompanyRepository companyRepository;
    private final CrmLeadTagRepository crmLeadTagRepository;
    private final ObjectMapper objectMapper;
    private final WhatsAppMessageRepository messageRepository;
    private final WhatsAppConversationRepository conversationRepository;
    private final MeetingRepository meetingRepository;
    private final WhatsAppBroadcastRecipientRepository broadcastRecipientRepository;
    private final FollowUpStatusRepository followUpStatusRepository;
    private final ChatMemoryService chatMemoryService;
    private final SupabaseStorageService supabaseStorageService;

    private static final Map<LeadStatus, String> STATUS_LABELS = createStatusLabels();

    private static Map<LeadStatus, String> createStatusLabels() {
        Map<LeadStatus, String> m = new HashMap<>();
        m.put(LeadStatus.NEW, "Novos Leads");
        m.put(LeadStatus.CONTACTED, "Em Contato");
        m.put(LeadStatus.QUALIFIED, "Qualificados");
        m.put(LeadStatus.MEETING_SCHEDULED, "Reunião");
        m.put(LeadStatus.PROPOSAL_SENT, "Proposta");
        m.put(LeadStatus.NEGOTIATION, "Negociação");
        m.put(LeadStatus.WON, "Ganhos");
        m.put(LeadStatus.LOST, "Perdidos");
        return m;
    }

    /**
     * Lista todos os leads da empresa
     */
    @Transactional(readOnly = true)
    public List<LeadResponse> getAllLeads(Company company) {
        return leadRepository.findAllByCompanyIdOrderByCreatedAtDescWithTags(company.getId())
                .stream()
                .map(this::toResponse)
                .collect(Collectors.toList());
    }

    /**
     * Tags cadastradas para a empresa (catálogo do CRM — sugestões + filtros).
     */
    @Transactional(readOnly = true)
    public List<LeadTagResponse> listCrmTagsForCompany(Company company) {
        return crmLeadTagRepository.findByCompany_IdOrderByNameAsc(company.getId()).stream()
                .map(t -> LeadTagResponse.builder().id(t.getId()).name(t.getName()).build())
                .collect(Collectors.toList());
    }

    /**
     * Lista leads paginados
     */
    @Transactional(readOnly = true)
    public Page<LeadResponse> getLeadsPaged(Company company, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"));
        return leadRepository.findByCompanyOrderByCreatedAtDesc(company, pageable)
                .map(this::toResponse);
    }

    /**
     * Busca leads por termo
     */
    @Transactional(readOnly = true)
    public Page<LeadResponse> searchLeads(Company company, String search, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"));
        return leadRepository.searchByCompany(company, search, pageable)
                .map(this::toResponse);
    }

    /**
     * Títulos personalizados das colunas do Kanban CRM (por empresa). Chaves = LeadStatus.name().
     */
    @Transactional(readOnly = true)
    public Map<String, String> getCrmKanbanColumnTitles(Company company) {
        return companyRepository.findById(company.getId())
                .map(this::parseCrmKanbanColumnTitles)
                .orElseGet(LinkedHashMap::new);
    }

    private Map<String, String> parseCrmKanbanColumnTitles(Company c) {
        String raw = c.getCrmKanbanColumnTitles();
        if (raw == null || raw.isBlank()) {
            return new LinkedHashMap<>();
        }
        try {
            Map<String, String> parsed = objectMapper.readValue(raw, new TypeReference<Map<String, String>>() {});
            Map<String, String> out = new LinkedHashMap<>();
            for (Map.Entry<String, String> e : parsed.entrySet()) {
                try {
                    LeadStatus st = LeadStatus.valueOf(e.getKey().trim().toUpperCase());
                    if (e.getValue() == null) {
                        continue;
                    }
                    String t = e.getValue().trim();
                    if (t.isEmpty()) {
                        continue;
                    }
                    if (t.length() > CRM_KANBAN_TITLE_MAX_LEN) {
                        t = t.substring(0, CRM_KANBAN_TITLE_MAX_LEN);
                    }
                    out.put(st.name(), t);
                } catch (IllegalArgumentException ignored) {
                    // ignora chave inválida
                }
            }
            return out;
        } catch (Exception e) {
            return new LinkedHashMap<>();
        }
    }

    /**
     * Persiste títulos customizados; remove entradas iguais ao padrão; mapa vazio limpa o JSON.
     */
    @Transactional
    public Map<String, String> updateCrmKanbanColumnTitles(Company company, Map<String, String> titles) {
        Company c = companyRepository.findById(company.getId())
                .orElseThrow(() -> new RuntimeException("Empresa não encontrada"));
        Map<String, String> cleaned = new LinkedHashMap<>();
        if (titles != null) {
            for (Map.Entry<String, String> e : titles.entrySet()) {
                try {
                    LeadStatus st = LeadStatus.valueOf(e.getKey().trim().toUpperCase());
                    if (e.getValue() == null) {
                        continue;
                    }
                    String t = e.getValue().trim();
                    if (t.isEmpty()) {
                        continue;
                    }
                    if (t.length() > CRM_KANBAN_TITLE_MAX_LEN) {
                        t = t.substring(0, CRM_KANBAN_TITLE_MAX_LEN);
                    }
                    String def = STATUS_LABELS.get(st);
                    if (def != null && t.equals(def)) {
                        continue;
                    }
                    cleaned.put(st.name(), t);
                } catch (IllegalArgumentException ignored) {
                    // ignora chave inválida
                }
            }
        }
        try {
            if (cleaned.isEmpty()) {
                c.setCrmKanbanColumnTitles(null);
            } else {
                c.setCrmKanbanColumnTitles(objectMapper.writeValueAsString(cleaned));
            }
            companyRepository.save(c);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Erro ao salvar títulos do CRM", e);
        }
        return cleaned;
    }

    /**
     * Busca lead por ID
     */
    @Transactional(readOnly = true)
    public LeadResponse getLeadById(Company company, UUID id) {
        Lead lead = leadRepository.findByIdAndCompanyWithTags(id, company.getId())
                .orElseThrow(() -> new RuntimeException("Lead não encontrado"));
        return toResponse(lead);
    }

    /**
     * Cria um novo lead
     */
    @Transactional
    public LeadResponse createLead(Company company, LeadRequest request) {
        Lead lead = Lead.builder()
                .company(company)
                .name(request.getName())
                .email(request.getEmail())
                .phone(request.getPhone())
                .status(request.getStatus() != null ? request.getStatus() : LeadStatus.NEW)
                .ownerName(request.getOwnerName())
                .notes(request.getNotes())
                .source(request.getSource())
                .trackId(request.getTrackId())
                .trackSource(request.getTrackSource())
                .utmSource(request.getUtmSource())
                .utmMedium(request.getUtmMedium())
                .utmCampaign(request.getUtmCampaign())
                .utmContent(request.getUtmContent())
                .utmTerm(request.getUtmTerm())
                .gclid(request.getGclid())
                .fbclid(request.getFbclid())
                .estimatedValue(request.getEstimatedValue())
                .leadScore(request.getLeadScore() != null ? request.getLeadScore() : 0)
                .build();

        lead = leadRepository.save(lead);
        if (request.getTags() != null) {
            syncLeadTags(company, lead, request.getTags());
            lead = leadRepository.save(lead);
        }

        Lead reloaded = leadRepository.findByIdAndCompanyWithTags(lead.getId(), company.getId())
                .orElse(lead);
        return toResponse(reloaded);
    }

    /**
     * Atualiza um lead existente
     */
    @Transactional
    public LeadResponse updateLead(Company company, UUID id, LeadRequest request) {
        Lead lead = leadRepository.findByIdAndCompany(id, company)
                .orElseThrow(() -> new RuntimeException("Lead não encontrado"));

        lead.setName(request.getName());
        lead.setEmail(request.getEmail());
        lead.setPhone(request.getPhone());
        if (request.getStatus() != null) {
            // Se o status foi alterado manualmente, marcar como qualificação manual
            if (!request.getStatus().equals(lead.getStatus())) {
                lead.setManuallyQualified(true);
            }
            lead.setStatus(request.getStatus());
        }
        lead.setOwnerName(request.getOwnerName());
        lead.setNotes(request.getNotes());
        lead.setSource(request.getSource());
        if (request.getEstimatedValue() != null) {
            lead.setEstimatedValue(request.getEstimatedValue());
        }
        if (request.getLeadScore() != null) {
            lead.setLeadScore(request.getLeadScore());
        }
        if (request.getTrackId() != null && !request.getTrackId().isBlank()) {
            lead.setTrackId(request.getTrackId().trim());
        }
        if (request.getTrackSource() != null && !request.getTrackSource().isBlank()) {
            lead.setTrackSource(request.getTrackSource().trim());
        }
        if (request.getUtmSource() != null && !request.getUtmSource().isBlank()) {
            lead.setUtmSource(request.getUtmSource().trim());
        }
        if (request.getUtmMedium() != null && !request.getUtmMedium().isBlank()) {
            lead.setUtmMedium(request.getUtmMedium().trim());
        }
        if (request.getUtmCampaign() != null && !request.getUtmCampaign().isBlank()) {
            lead.setUtmCampaign(request.getUtmCampaign().trim());
        }
        if (request.getUtmContent() != null && !request.getUtmContent().isBlank()) {
            lead.setUtmContent(request.getUtmContent().trim());
        }
        if (request.getUtmTerm() != null && !request.getUtmTerm().isBlank()) {
            lead.setUtmTerm(request.getUtmTerm().trim());
        }
        if (request.getGclid() != null && !request.getGclid().isBlank()) {
            lead.setGclid(request.getGclid().trim());
        }
        if (request.getFbclid() != null && !request.getFbclid().isBlank()) {
            lead.setFbclid(request.getFbclid().trim());
        }

        if (request.getTags() != null) {
            syncLeadTags(company, lead, request.getTags());
        }

        lead = leadRepository.save(lead);
        Lead reloaded = leadRepository.findByIdAndCompanyWithTags(lead.getId(), company.getId())
                .orElse(lead);
        return toResponse(reloaded);
    }

    private void syncLeadTags(Company company, Lead lead, List<String> rawLabels) {
        LinkedHashMap<String, String> byLower = new LinkedHashMap<>();
        for (String raw : rawLabels) {
            if (raw == null) {
                continue;
            }
            String tr = raw.trim();
            if (tr.isEmpty()) {
                continue;
            }
            byLower.putIfAbsent(tr.toLowerCase(Locale.ROOT), tr);
        }
        Set<CrmLeadTag> resolved = new LinkedHashSet<>();
        for (String label : byLower.values()) {
            CrmLeadTag tag = crmLeadTagRepository
                    .findByCompanyIdAndNameNormalized(company.getId(), label)
                    .orElseGet(() -> crmLeadTagRepository.save(CrmLeadTag.builder()
                            .company(company)
                            .name(label)
                            .build()));
            resolved.add(tag);
        }
        lead.getCrmTags().clear();
        lead.getCrmTags().addAll(resolved);
    }

    /**
     * Deleta um lead
     */
    @Transactional
    public void deleteLead(Company company, UUID id) {
        Lead lead = leadRepository.findByIdAndCompany(id, company)
                .orElseThrow(() -> new RuntimeException("Lead não encontrado"));

        chatMemoryService.clearHistory(id.toString());

        purgeWhatsAppThreadForLead(company, id);

        meetingRepository.clearLeadReference(id);
        broadcastRecipientRepository.clearLeadReference(id);

        if (lead.getProfilePictureUrl() != null && !lead.getProfilePictureUrl().isEmpty()) {
            supabaseStorageService.tryDeletePublicStorageObject(lead.getProfilePictureUrl());
        }

        leadRepository.deleteById(id);
    }

    /**
     * Remove completamente o atendimento WhatsApp do lead: follow-up, mensagens, mídia no Storage
     * e a conversa (some da lista).
     */
    private void purgeWhatsAppThreadForLead(Company company, UUID leadId) {
        Set<UUID> convIds = new LinkedHashSet<>();
        for (WhatsAppConversation c : conversationRepository.findByLead_Id(leadId)) {
            if (c.getCompany() != null && Objects.equals(c.getCompany().getId(), company.getId())) {
                convIds.add(c.getId());
            }
        }
        for (UUID cid : messageRepository.findDistinctConversationIdsByLeadId(leadId)) {
            conversationRepository.findById(cid).ifPresent(conv -> {
                if (conv.getCompany() != null && Objects.equals(conv.getCompany().getId(), company.getId())) {
                    convIds.add(cid);
                }
            });
        }
        for (UUID cid : convIds) {
            WhatsAppConversation conv = conversationRepository.findById(cid).orElse(null);
            if (conv == null) {
                continue;
            }
            for (String mediaUrl : messageRepository.findMediaUrlsByConversationId(conv.getId())) {
                supabaseStorageService.tryDeletePublicStorageObject(mediaUrl);
            }
            if (conv.getProfilePictureUrl() != null && !conv.getProfilePictureUrl().isEmpty()) {
                supabaseStorageService.tryDeletePublicStorageObject(conv.getProfilePictureUrl());
            }
            messageRepository.deleteByConversation(conv);
            followUpStatusRepository.findByConversationId(conv.getId()).ifPresent(followUpStatusRepository::delete);
            chatMemoryService.clearHistory(conv.getId().toString());
            conversationRepository.delete(conv);
        }
        messageRepository.deleteByLead_Id(leadId);
    }

    /**
     * Converte entidade para DTO
     */
    private LeadResponse toResponse(Lead lead) {
        List<LeadTagResponse> tagDtos = lead.getCrmTags() == null ? List.of() : lead.getCrmTags().stream()
                .map(t -> LeadTagResponse.builder()
                        .id(t.getId())
                        .name(t.getName())
                        .build())
                .sorted(java.util.Comparator.comparing(LeadTagResponse::getName, String.CASE_INSENSITIVE_ORDER))
                .collect(Collectors.toList());

        return LeadResponse.builder()
                .id(lead.getId())
                .name(lead.getName())
                .email(lead.getEmail())
                .phone(lead.getPhone())
                .status(lead.getStatus().name())
                .statusLabel(STATUS_LABELS.getOrDefault(lead.getStatus(), lead.getStatus().name()))
                .ownerName(lead.getOwnerName())
                .notes(lead.getNotes())
                .source(lead.getSource())
                .trackId(lead.getTrackId())
                .trackSource(lead.getTrackSource())
                .utmSource(lead.getUtmSource())
                .utmMedium(lead.getUtmMedium())
                .utmCampaign(lead.getUtmCampaign())
                .utmContent(lead.getUtmContent())
                .utmTerm(lead.getUtmTerm())
                .gclid(lead.getGclid())
                .fbclid(lead.getFbclid())
                .estimatedValue(lead.getEstimatedValue())
                .leadScore(lead.getLeadScore() != null ? lead.getLeadScore() : 0)
                .profilePictureUrl(lead.getProfilePictureUrl())
                .createdAt(lead.getCreatedAt())
                .updatedAt(lead.getUpdatedAt())
                .tags(tagDtos)
                .build();
    }
}
